lib/matrix.gi

```diff
@@ -2440,9 +2456,16 @@ InstallMethod( SemiEchelonMatTransformation,
     return SemiEchelonMatTransformationDestructive( copymat );
 end);
 
-InstallMethod( SemiEchelonMatTransformationDestructive,
+InstallOtherMethod(SemiEchelonMatTransformation,"matrix objects",[IsMatrixObj],
+function( mat )
+  local copymat;
+  copymat:=MutableCopyMatrix(mat);
+  return SemiEchelonMatTransformationDestructive( copymat );
+end);
+
+InstallOtherMethod( SemiEchelonMatTransformationDestructive,
     "generic method for matrices",
-    [ IsMatrix and IsMutable],
+    [ IsMatrixOrMatrixObj and IsMutable],
```

Comment:
This is not appropriate: this method requires `mat` to be in filter `IsRowListMatrix`, i.e., to be  row-based and allow access to matrix objects via `mat[i]`. It also uses `DefaultFieldOfMatrix`, and indeed, it expects non-zero entries to be invertible.

Reply:
In general `mat[i]` won't work on general MatrixObj instances. We debated about providing a default implementation which uses proxy objects to "fake it", but there are all kinds of pitfalls with this, and results in super inefficient code. So in the end, we agreed that it's better to get an error and be able to fix the affected code.

Ideally, it shouldn't be necessary to have two versions of the code, though: if `AddMatrixRows` etc. are used then the resulting code should work both for new and old style matrices, i.e. for everything in the filter `IsMatrixOrMatrixObj`. The main question is performance: using `AddMatrixRows` might be slower initially when used on "old style matrices". However, there is still quite some optimization potential there. E.g. one source of slow down is method dispatch is slow for lists, and thus for old-style matrices; this could be dealt with using `InstallEarlyMethod` and/or some kernel methods; I'd be happy to help work on tuning this.